proposals/low-level-struct-improvements.md

```diff
@@ -0,0 +1,972 @@
+lifetime of the return value of these methods is never restricted by the inputs.
+The existing span safety rules consider such values as effectively always 
+*safe-to-escape* outside the enclosing method.
+
+```cs
+class ExistingScenarios
+{
+    Span<T> CreateSpan<T>(ref T p)
+    {
+        // The implementation of this method is irrelevant. From the point of
+        // the consumer the returned value is always safe to return.
+        ... 
+    }
+
+    Span<T> Examples<T>(ref T p, T[] array)
+    {
+        // Legal today
+        return CreateSpan(ref p); 
+
+        // Legal today, must remain legal
+        T local = default;
+        return CreateSpan(ref local);
+
+        // Legal for any possible value that could be used as an argument
+        return CreateSpan(...);
+    }
+}
+```
+
+The reason that all of the above samples are legal is because in the 
+existing design there is no way for the return `Span<T>` to store a reference 
+to the input state of the method call. This is because the span safety rules 
+explicitly depend on `Span<T>` not having a constructor which takes a `ref` 
+parameter and stores it as a field. 
+
+```cs
+class ExistingAssumptions
+{
+    Span<T> CreateSpan<T>(ref T p)
+    {
+        // !!! Cannot happen today !!!
+        // The existing span safety rules specifically call out that this method
+        // cannot exist hence they can assume all returns from CreateSpan are
+        // safe to return.
+        return new Span<T>(ref p);
+    }
+}
+```
+
+The rules we define for `ref` fields must ensure the `Span<T>` constructor 
+properly restricts the *safe-to-escape* scope of constructed objects in the 
+cases it captures `ref` state. At the same time it must ensure that we don't 
+break the existing consumption rules for methods like `CreateSpan<T>`. 
+
+```cs
+class GoalOfRefFields
+{
+    Span<T> CreateSpan<T>(ref T p)
+    {
+        // ERROR: the existing consumption rules for CreateSpan believe this 
+        // can never happen hence we must continue to enforce that it cannot
+        return new Span<T>(ref p);
+
+        // Okay: this is legal today
+        return new Span<int>(new int[] { });
+    }
+
+    Span<int> ConsumptionCompatibility()
+    {
+        // Okay: this is legal today and must remain legal.
+        int local = 42;
+        return CreateSpan(ref local);
+
+        // Okay: the arguments don't actually matter here. Literally any value 
+        // could be passed to this method and the return of it would still be 
+        // *safe-to-escape* outside the enclosing method. 
+        return CreateSpan(...);
+    }
+}
+```
+
+This tension between allowing constructors such as `Span<T>(ref T field)` and 
+ensuring compatibility with `ref struct` returning methods like `CreateSpan<T>` 
+is a key pivot point in the design of `ref` fields.
+
+To do this we will change the escape rules for a constructor invocation, which
```

Comment:
Why not instead of changing the rules for ref struct constructors (which is a breaking change), mark them with [RefEscapes] explicitly, and consider `this` to be the return type of a constructor?

Reply:
These rules only apply to constructors on `ref struct` which directly contain `ref` fields. Given that `ref` fields don't exist today this can't be a breaking change because there are no constructors that they presently apply to.

Reply:
Except for `Span<T>`, right?